### PR TITLE
TypeScript and docs

### PR DESCRIPTION
**Pull request by Thomaash**
Solves #1 and #2.

- Types are not perfect (quite a few anies in the declarations), I left some commented types there in case someone will find the bug. However most types are good.
- The tests are unchanged and pass.
- Tested in vis-chart, examples seem to be working okay (tests fail event with vis-util from npm).
- Also, I replaced some custom implementations with standard ECMA script (Object.assign, Object.values, {...o}, …), Babel will handle compatibility.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,41 +1,73 @@
...
+  "bugs": {
+    "url": "https://github.com/visjs/vis-util/issues"
+  },
+  "homepage": "https://github.com/visjs/vis-util",
+  "files": [
+    "dist",
+    "LICENSE*"
+  ],
   "dependencies": {
+    "moment": "^2.24.0",
```
Comment (mojoaxel) [inline package.json]: :+1:

Comment (Thomaash) [reply package.json]: Thinking about this, shouldn't Moment be a peer dependency?

Comment (mojoaxel) [reply package.json]: You are write! I'm not sure how to handle this at the moment. For the browser version we definitely need to generate a bundle without moment.
Feel free to provide a solution for this. For now on I'm also ok with bundling moment.js by defrault and remove it in a later version...not ideal..

Comment (Thomaash) [reply package.json]: I'll look into this.

Comment (mojoaxel) [reply package.json]: Here is how vis used to handle it: https://github.com/visjs/vis_original/blob/master/lib/module/moment.js

Comment (Thomaash) [reply package.json]: But then it always has to be bundled, it only remains unused if something else already imported Moment.

Comment (Thomaash) [reply package.json]: Peer dependency would be more along the lines “shipping this is your job, not ours”.

Comment (Thomaash) [reply package.json]: Maybe multiple builds? With and without Moment, so that users can pick which better suits their needs?

Comment (Thomaash) [reply package.json]: We can create builds without peer dependencies (smaller files for people who install through npm or ship their own) and with dependencies (bigger files for people who just want it to work) like this:
```javascript
import babel from 'rollup-plugin-babel'
import resolve from 'rollup-plugin-node-resolve'
import packageJSON from './package.json'

// This is necessary for Moment to work.
import commonjs from 'rollup-plugin-commonjs'

const external = Object.keys(packageJSON.peerDependencies || {})

const plugins = [
  resolve({
    extensions: ['.ts', '.js', '.json'],
  }),
  commonjs(),
  babel({
    extensions: ['.ts', '.js'],
  }),
]

export default [
  {
    input: 'src/index.ts',
    output: {
      file: 'dist/esm.js',
      format: 'esm',
    },
    external,
    plugins,
  },
  {
    input: 'src/index.ts',
    output: {
      file: 'dist/esm.with-deps.js',
      format: 'esm',
    },
    plugins,
  },
  {
    input: 'src/index.ts',
    output: {
      file: 'dist/umd.js',
      format: 'umd',
      name: 'visUtil',
    },
    external,
    plugins,
  },
  {
    input: 'src/index.ts',
    output: {
      file: 'dist/umd.with-deps.js',
      format: 'umd',
      name: 'visUtil',
    },
    plugins,
  },
]
```

Comment (mojoaxel) [reply package.json]: Look interesting.
This is something we can discuss further on the main packeges: vis-timeline, vis-network and vis-graph3d. 
For now we could leave it as it is!

Comment (Thomaash) [reply package.json]: Okay.

File: rollup.config.js [hunk 2]
```diff
@@ -15,15 +18,15 @@ export default [
...
-    plugins: [resolve(resolveConfig), babel(babelConfingBase)],
+    plugins: [resolve(resolveConfig), commonjs(), babel(babelConfingBase)],
   },
   {
     input: 'src/index.ts',
     output: {
       file: 'dist/umd.js',
       format: 'umd',
-      name: 'visUUID',
+      name: 'visUtil',
```
Comment (mojoaxel) [inline rollup.config.js]: The name should be just `vis`.

**Pull request merged**